### DISCUSSION TITLE
z fold bindings (text editor)

### DISCUSSION THREAD

**Discussion by martin-braun** [vscode-neovim/vscode-neovim#1302] [Ideas]
Should we target Neovim z folding support or fallback to VSCode line folding that gets triggered by z bindings?

Benefits of Neovim's z folding is: 

- foldmethods marker support, thus compatible folds in Neovim and VSCode
- richer controls, Neovim offers more features on foldings, like nested folds
- consistent behavior across all editors

Benefits of VSCode:

- Easier to implement, the Neovim folds are very broke in the buffer atm
- Mouse support
- Visually more pleasing

**Comment by theol0403** [comment]
We can't go with neovim's folding as is, because it will break LSP in the vscode buffer. The correct solution would be:

1. vscode folds with custom bindings, and fix cursor movement (skip over folds) so that the experience is good
2. fully sync neovim and vscode folds

**Reply by martin-braun** [reply]
Skipping over a VS code fold manually seem to work: 

![2023-07-11 20 20 13](https://github.com/vscode-neovim/vscode-neovim/assets/24259245/38e9a5a8-82a4-4ef4-b8e0-4154fa0b0ad3)

So I guess Neovim sees the full buffer as if there were no folds. The bad part is that once you enter the folded area, it automatically unfolds.

Also, wouldn't (2.) also break LSPs?

**Reply by theol0403** [reply]
Yeah, that's what https://github.com/vscode-neovim/vscode-neovim/pull/502 is supposed to do. Haven't had the time to merge it, because it needs some good bindings too.

I think when neovim folds are made, it won't actually delete the buffer lines. I'm not entirely sure, but I think folds are a visual element as opposed to a buffer element. Not sure what the APIs are, but I think the buffer would be untouched and we would just have to forward fold/expand to let vscode know.
